wycheproof/src/main.rs

```diff
@@ -0,0 +1,119 @@
+//! Tool to convert Wycheproof test vectors to raw hex format
+use maplit::btreemap;
+use std::collections::BTreeMap;
+use std::io::Write;
+use structopt::StructOpt;
+
+mod aead;
+mod aes_cmac;
+mod aes_siv;
+mod ecdsa;
+mod ed25519;
+mod hkdf;
+mod prf;
+mod wycheproof;
+
+#[derive(Debug, StructOpt)]
+struct Opts {
+    #[structopt(
+        long,
+        help = "The location of the Wycheproof repository",
+        default_value = ""
+    )]
+    wycheproof_dir: String,
+
+    #[structopt(
+        long,
+        help = "Algorithm to convert to raw hex format",
+        default_value = ""
+    )]
+    algorithm: String,
+
+    #[structopt(long, help = "Destination file", default_value = "")]
+    out: String,
+
+    #[structopt(
+        long,
+        help = "Destination file for test descriptions (one per line)",
+        default_value = ""
+    )]
+    txt: String,
+}
+
+/// Test information
+pub struct TestInfo {
+    /// Raw data for the tests.
+    pub data: Vec<Vec<u8>>,
+    /// Test case description.
+    pub desc: String,
+}
+
+/// Generator function which takes input parameters:
+///  - contents of Wycheproof test data file
+///  - algorithm name
+/// and returns the raw contents, together  with a list of test identifiers (one per entry).
+type BlbGenerator = fn(&[u8], &str) -> Vec<TestInfo>;
+
+struct Algorithm {
+    pub file: &'static str,
+    pub generator: BlbGenerator,
+}
+
+fn main() {
+    let opts = Opts::from_args();
+    if opts.wycheproof_dir.is_empty() {
+        panic!("Need location of wycheproof data");
+    }
+
+    let algorithms: BTreeMap<String, Algorithm> = btreemap! {
+        "AES-GCM".to_string() => Algorithm{file: "aes_gcm_test.json", generator: aead::aes_gcm_generator},
+        "AES-GCM-SIV".to_string() => Algorithm{file: "aes_gcm_siv_test.json", generator: aead::aes_gcm_generator},
+        "CHACHA20-POLY1305".to_string() => Algorithm{file: "chacha20_poly1305_test.json", generator: aead::chacha20_poly1305},
```

Comment:
A simple `match` should do just fine here, no? Yes, unfortunately compiler is [not smart enough](https://rust.godbolt.org/z/f1q47h) to join common prefixes, so it will generate a series of "dumb" comparisons, but this part is not performance-sensitive, so I don't think it's worth to bother with `BTreeMap` here.

Reply:
Using a map allows the error message for an unrecognised algorithm name (l.92) to include the (sorted) list of valid names, but I guess that won't end up being that helpful (use of the tool is likely to be in a script or done by people who can just read the source).

Reply:
Yeah, as I've said earlier, this crate is for our internal usage, so a bit hard to understand runtime errors should be fine. Alternatively you can use this approach:
```rust
const ALG_MAP: &[(&str, Algorithm)] = &[
    ("AES-GCM", Algorithm{file: "aes_gcm_test.json", generator: aead::aes_gcm_generator}),
    ("AES-GCM-SIV".to_string(), Algorithm{file: "aes_gcm_siv_test.json", generator: aead::aes_gcm_generator}),
    // ...
];

let alg = ALG_MAP
    .iter()
    .find(|&(alg_name, _)| name = algorithm)
    .map(|&(_, alg) alg);
let alg = match alg {
    Some(alg) => alg,
    None => {
        // print error with list of valid names and exit
    }
}
```